src/librustc_codegen_llvm/builder.rs

```diff
@@ -56,6 +56,99 @@ bitflags! {
         const NONTEMPORAL = 1 << 1;
     }
 }
+/// Check that we can actually cast between these addr spaces.
+pub fn check_addr_space_cast<T>(cx: &CodegenCx, val: ValueRef, dest: &T)
+    where T: AddrSpaceOf,
+{
+    let src_ty = val_ty(val);
+
+    debug_assert!(src_ty.is_ptr());
+
+    let left = src_ty.address_space();
+    let right = dest.addr_space().expect("dest has no address space?");
+    if !cx.can_cast_addr_space(left, right) {
+        let msg = format!("Target incompatible address space cast:\n\
+                           source addr space `{}`, dest addr space `{}`",
+                          left, right);
+        if llvm::dumping_enabled() {
+            unsafe {
+                println!("{}", msg);
+                llvm::LLVMDumpValue(val);
+                println!();
+                println!("source ty:");
+                llvm::LLVMDumpType(src_ty.to_ref());
```

Comment:
Wait, you don't need this, we already have printing for both values and types. It's `{:?}` with just `src_ty` and with `Value(val)` (a wrapper hack, @irinagpopa already has it removed on her branch, so `val` would work directly).
That is, you can put the value and the type in the message, *always*, and you don't need any of the changes to the build system.

Reply:
Oh, nice! I'll get that changed